portal/common/lib/rpc_selector.ts

```diff
@@ -0,0 +1,159 @@
+            timeoutPromise,
+        ]);
+
+        if (this.isValidResponse(result)) {
+            return result;
+        } else {
+            console.error("Invalid response from selected client");
+            throw new Error("Invalid response from selected client");
+        }
+    }
+
+    // Fallback to querying all clients using Promise.any.
+    private async callFallbackClients<T>(methodName: string, args: any[]): Promise<T> {
+        const clientPromises = this.clients.map((client) =>
+            new Promise<{ result: T; client: SuiClient }>(async (resolve, reject) => {
+                try {
+                    const method = (client as any)[methodName] as Function;
+                    if (!method) {
+                        reject(new Error(`Method ${methodName} not found on client`));
+                        return;
+                    }
+                    const result = await method.apply(client, args);
+                    if (this.isValidResponse(result)) {
+                        resolve({ result, client });
+                    } else {
+                        reject(new Error("Invalid response"));
+                    }
+                } catch (error: any) {
+                    reject(error);
+                }
+            }),
+        );
+
+        try {
+            const { result, client } = await Promise.any(clientPromises);
+            // Update the selected client for future calls.
+            this.selectedClient = client;
+            return result;
+        } catch {
+            throw new Error(`Failed to contact fallback RPC clients.`);
+        }
+    }
+
+    private isValidResponse(result: SuiObjectResponse | SuiObjectResponse[] | string): boolean {
+        // GetObject or getDynamicFieldObject
+        if (result == null) {
+            return false;
+        }
+
+        // SuiNS
+        if (typeof result === 'string') {
+            return result.trim().length > 0;
+        }
+
+        // MultiGetObject
+        if (Array.isArray(result)) {
+            return result.some((item) => item != null);
+        }
+        return true;
+    }
+
+    public async getObject(input: GetObjectParams): Promise<SuiObjectResponse> {
+        return this.invokeWithFailover<SuiObjectResponse>("getObject", [input]);
+    }
+
+    public async multiGetObjects(
+        input: MultiGetObjectsParams,
+    ): Promise<SuiObjectResponse[]> {
+        return this.invokeWithFailover<SuiObjectResponse[]>("multiGetObjects", [input]);
+    }
+
+    public async getDynamicFieldObject(
+        input: GetDynamicFieldObjectParams,
+    ): Promise<SuiObjectResponse> {
+        return this.invokeWithFailover<SuiObjectResponse>("getDynamicFieldObject", [
+            input,
+        ]);
+    }
+
+    public async call<T>(method: string, args: any[]): Promise<T> {
+        return this.invokeWithFailover<T>(method, args);
+    }
+}
+
+const rpcSelectorSingleton = RPCSelector.getInstance();
+export default rpcSelectorSingleton;
```

Comment:
Remark: in general, i'm not a huge fan of singletons, and prefer dependency injection whenever possible. Yet, this is good enough for now imo.